### ISSUE TITLE
[ET-VK] Embedding with vocab > 16384 silently returns wrong values on the texture path; force_fp16 breaks all-MiniLM-L6-v2

### ISSUE THREAD

**Issue by msluszniak** [pytorch/executorch#22333]
### 🐛 Describe the bug

`aten.embedding` returns wrong values on Vulkan when the vocabulary exceeds 16384 rows and the output is a texture. There is no error; the model produces well-formed but semantically wrong embeddings.

This breaks `sentence-transformers/all-MiniLM-L6-v2` (vocab 30522) under `force_fp16`.

### Threshold

`nn.Embedding(V, 384)`, 64 random indices, cosine of the Vulkan output against the CPU reference on an Adreno 840:

| vocab | fp32 | fp16 |
| --- | --- | --- |
| 2048 | 1.000000 | 1.000000 |
| 4096 | 1.000000 | 1.000000 |
| 8192 | 1.000000 | 1.000000 |
| **16384** | **1.000000** | **1.000000** |
| **16385** | **0.255916** | **0.255916** |
| 20000 | 0.256048 | 0.256048 |
| 30522 | 0.259350 | 0.259349 |

The break is exactly at 16384, which is this device's `maxImageDimension2D`, and it is precision-independent. All of these dispatch `embedding_texture3d_{float,half}`.

### Why force_fp16 turns this into a real-model failure

`all-MiniLM-L6-v2` at the published 254-token shape, 8 sentences, embeddings compared against the CPU reference:

| | mean cosine | pairwise similarity max abs diff | top-1 nearest neighbour preserved |
| --- | --- | --- | --- |
| fp32 | 0.999999 | 0.00031 | 8 / 8 |
| **fp16** | **0.219010** | **0.68950** | **1 / 8** |

One embedding comes back at cosine -0.016 against its reference. The semantic structure is destroyed, so this silently breaks retrieval rather than merely degrading it.

The two differ only in which embedding kernel they reach:

```
fp32:  embedding_buffer_float
fp16:  embedding_texture3d_half
```

`force_fp16` biases the graph toward texture storage (`TagMemoryMetaPass.constrain_op_arg_repset` calls `try_constrain_with_arg_repset(arg_i, utils.ANY_TEXTURE)` unconditionally when `force_fp16` is set), which moves the embedding output from buffer to texture and onto the broken path. fp32 escapes only by landing on the buffer kernel.

### What I ruled out

For the MiniLM failure specifically: dynamic shapes (a static export is bit-identical), the padding and attention mask (a fully unmasked 254-token input fails the same, cosine 0.169), and the mean-pool/normalise tail (the pre-pool token output is already wrong). Exporting only `model.embeddings` reproduces it at cosine 0.33, before any encoder layer.

For the isolated case: it is not the legacy texture-weight path. `embedding()` in `Embedding.cpp` prepacks the weight as `kBuffer` for these models, and the dispatched kernel is the non-legacy `embedding_texture3d_*`. I tried guarding the legacy branch on `max_texture2d_dim()` and it changed nothing, confirming that branch is not involved.

I did not find the exact mechanism inside `embedding_texture.glsl`. `load_weight_texel()` computes `embedding_idx * width(weight) + dim_idx`, which does not overflow int32 at these sizes, so the 16384 boundary most likely comes from how the indices or the weight are addressed rather than from that multiply.

### Suggested direction

Two things seem worth separating:

1. `force_fp16` should not push a tensor toward texture storage without consulting texture limits, so oversized cases keep the working buffer kernel.
2. Exceeding texture extents should fail loudly rather than silently returning garbage.

### Repro

Scripts are straightforward to reconstruct from the table above: export `nn.Embedding(V, 384)` with `VulkanPartitioner()` for V on either side of 16384 and compare against the CPU reference.

### Versions

ExecuTorch 1.4.1 for the export, runtime at c27baa8031. Device: Samsung Galaxy S26 Ultra, Snapdragon SM8850, Adreno 840, Android 16.

cc @SS-JIA @manuelcandales @digantdesai @cbilgin @mergennachin @kimishpatel @iseeyuan

**Comment by msluszniak** [comment]
Some follow-up on this: the addressing law, two more affected models, and an AOT-only workaround.

## The corrupted reads follow an exact law

`nn.Embedding(49408, 512)`, 16 random indices, Adreno 840, comparing each output row against the flattened weight to find where it actually came from:

```
row 0: want flat 14637056 (id 28588)  got 14647312   (+10256)
row 1: want flat  3563008 (id  6959)  got 14647313   (row 0's base +1)
row 2: want flat 10021376 (id 19573)  got 14647314   (+2)
row 3: want flat 20807680 (id 40640)  got 14647315   (+3)
row 4: want flat  9209344 (id 17987)  got  9219600   (+10256)
row 5: want flat 25163264 (id 49147)  got  9219601   (row 4's base +1)
```

Output rows group in fours. Within each group all four lanes read from a single index, the group's first, and then step by one element. The `+10256` base offset is constant across groups and independent of the index.

## That is `load_weight_texel`

```glsl
VEC4_T load_weight_texel(const int embedding_idx, const int dim_idx) {
  int buf_i = embedding_idx * int(width(weight)) + dim_idx;
  [[unroll]] for (int i = 0; i < 4; ++i) {
    weight_texel[i] = T(t_weight[buf_i++]);
  }
}
```

It fills the texel with four consecutive elements along the weight's width, which is correct only if the output texture is width-packed, and `main()` calls `load_embedding_idx` once per texel. When the output is packed along the token dim instead, the four components of a texel need four different rows of the weight. The shader never consults `out_layout`'s packed dim.

Still unexplained: the constant `+10256`, and why 16384 is where the behaviour switches. Those are consistent with a layout choice that flips at the texture limit rather than with the shader itself, but I have not confirmed that.

## Two more models hit it

Both under `force_fp16`, cosine against the CPU reference, on Adreno 840 and Mali-G76:

| model | vocab | Adreno | Mali |
| --- | --- | --- | --- |
| CLIP ViT-B/32 text | 49408 | 0.511837 | 0.510685 |
| all-mpnet-base-v2 | 30527 | 0.077958 | 0.075909 |

CLIP ViT-B/32 **vision** is unaffected at cos 0.99997, which fits the diagnosis: it has no token embedding, only a conv patch embed and a 50 row position embedding.

## AOT-only workaround

`_passes/tag_memory_meta_pass.py`, `constrain_op_arg_repset`:

```python
if self.force_fp16:
    op_repsets.try_constrain_with_arg_repset(arg_i, utils.ANY_TEXTURE)
```

is unconditional, which is what moves the embedding output onto the texture path. Skipping it when the op node, or the arg node at `arg_i`, is an `aten.embedding` keeps the embedding on the correct `embedding_buffer` kernel while the rest of the graph still goes fp16 and texture.

CLIP text fp16 goes from cos 0.511837 to **0.999999** with that change alone. It is export side only, so it needs no runtime rebuild, which makes it a usable stopgap for anyone blocked on this today.

One caveat worth recording: it is not free. Pinning the embedding to buffer storage pulls the surrounding prologue off texture too, and the patched fp16 CLIP text then runs 113 to 119 ms on Mali-G76 against 85 to 89 ms for plain fp32. So it recovers correctness and the fp16 size saving, but not a speed win over fp32.

This supports suggestion 1 in the original report: `force_fp16` should consult texture limits rather than constraining to `ANY_TEXTURE` unconditionally.

Runtime at 23ed765a51 (1.4.1 plus local reduce/softmax dispatch fixes), export on 1.4.1. Devices: Galaxy S26 Ultra (Adreno 840) and Galaxy S10+ (Mali-G76).
